opm/simulators/linalg/DILU.hpp

```diff
@@ -183,8 +157,198 @@ class SeqDilu : public PreconditionerWithUpdate<X, Y>
 private:
     //! \brief The matrix we operate on.
     const M& A_;
+    //! \brief Copy of A_ that is reordered to store rows that can be computed simultaneously next to each other to
+    //! increase cache usage when multithreading
+    M A_reordered_;
     //! \brief The inverse of the diagnal matrix
     std::vector<typename M::block_type> Dinv_;
+    // //! \brief Stores the Dinv_ indexed according to the rows sorted by level
+    // std::vector<typename M::block_type> Dinv_;
+    //! \brief SparseTable storing each row by level
+    Opm::SparseTable<size_t> level_sets;
+    //! \brief converts from index in reordered structure to index natural ordered structure
+    std::vector<size_t> reordered_to_natural_;
+    //! \brief converts from index in natural ordered structure to index reordered strucutre
+    std::vector<size_t> natural_to_reorder_;
+    //! \brief Boolean value describing whether or not to use multithreaded version of functions
+    bool use_multithreading;
+
+    void serialUpdate()
+    {
+        for (size_t row = 0; row < A_.N(); ++row) {
+            Dinv_[row] = A_[row][row];
+        }
+        for (auto row = A_.begin(); row != A_.end(); ++row) {
+            const auto row_i = row.index();
+            auto Dinv_temp = Dinv_[row_i];
+            for (auto a_ij = row->begin(); a_ij.index() < row_i; ++a_ij) {
+                const auto col_j = a_ij.index();
+                const auto a_ji = A_[col_j].find(row_i);
+                // if A[i, j] != 0 and A[j, i] != 0
+                if (a_ji != A_[col_j].end()) {
+                    // Dinv_temp -= A[i, j] * d[j] * A[j, i]
+                    Dinv_temp -= (*a_ij) * Dune::FieldMatrix(Dinv_[col_j]) * (*a_ji);
+                }
+            }
+            Dinv_temp.invert();
+            Dinv_[row_i] = Dinv_temp;
+        }
+    }
+
+    void parallelUpdate()
+    {
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+        for (size_t row = 0; row != A_.N(); ++row) {
+            Dinv_[natural_to_reorder_[row]] = A_[row][row];
+        }
+
+        // TODO: is there a better/faster way of copying all values?
```

Comment:
Even if there is no way to perform a single memory copy that moves all the values, it would still be possible to OpenMP parallelize this section as every write is to a distinct memory location